domains/client/domain-executor/src/tests.rs

```diff
@@ -256,55 +282,48 @@ async fn set_new_code_should_work() {
     let tokio_handle = tokio::runtime::Handle::current();
 
     // Start Ferdie
-    let (ferdie, ferdie_network_starter) = run_primary_chain_validator_node(
+    let mut ferdie = MockPrimaryNode::run_mock_primary_node(
         tokio_handle.clone(),
         Ferdie,
-        vec![],
         BasePath::new(directory.path().join("ferdie")),
-    )
-    .await;
-    ferdie_network_starter.start_network();
+    );
 
     // Run Alice (a system domain authority node)
     let alice = domain_test_service::SystemDomainNodeBuilder::new(
         tokio_handle.clone(),
         Alice,
         BasePath::new(directory.path().join("alice")),
     )
-    .connect_to_primary_chain_node(&ferdie)
-    .build(Role::Authority, false, false)
+    .build_with_mock_primary_node(Role::Authority, &mut ferdie)
     .await;
 
-    ferdie.wait_for_blocks(1).await;
-
-    let new_runtime_wasm_blob = b"new_runtime_wasm_blob".to_vec();
+    futures::join!(alice.wait_for_blocks(1), ferdie.produce_blocks(1))
+        .1
+        .unwrap();
 
-    let best_number = alice.client.info().best_number;
-    let primary_number = best_number + 1;
-    // Although we're processing the bundle manually, the original bundle processor still works in
-    // the meanwhile, it's possible the executor alice already processed this primary block, expecting next
-    // primary block, in which case we use a dummy primary hash instead.
-    //
-    // Nice to disable the built-in bundle processor and have a full control of the executor block
-    // production manually.
-    let primary_hash = ferdie
-        .client
-        .hash(primary_number)
-        .unwrap()
-        .unwrap_or_else(Hash::random);
-    alice
-        .executor
-        .clone()
-        .process_bundles((primary_hash, primary_number))
-        .await;
+    // Trigger a `RuntimeEnvironmentUpdated` digest in the primary chain, use `set_code` will
+    // fail due to `InvalidTransaction::ExhaustsResources` error thus use `set_heap_pages` instead
+    let set_heap_pages_tx = subspace_test_service::construct_extrinsic(
+        ferdie.client.as_ref(),
+        pallet_sudo::Call::sudo {
```

Comment:
There is an API that skips the runtime code check right below https://github.com/paritytech/substrate/blob/dd45632f298db02f6946c906296bbf86a74f4f46/frame/system/src/lib.rs#L414.

However, in order to extract the system domain code from the new primary runtime, we still need it to be a _proper primary runtime code_ with an arbitrary system domain runtime code included. I'm not sure the requirement of the minimal proper primary runtime code, but anyway, you don't have to use `set_heap_pages` here.

Reply:
Seems we do need a proper primary runtime code even for this test, otherwise, the executor will shut down and the test will fail due to:
```
2023-04-04 17:03:03 [//Alice (SystemDomain)] Cannot create a runtime error=Other("cannot deserialize module: InvalidMagic")
2023-04-04 17:03:03 [//Alice (SystemDomain)] Failed to process primary block error=Application(VersionInvalid("cannot deserialize module: InvalidMagic"))
```

Reply:
Hmm, I have some thoughts on creating a minimal primary runtime code, but unsure if that works actually. Feel free to revert this test or mark it as incomplete because this PR does not actually cover the new domain runtime are updated accordingly upon the new primary runtime so that we can push this PR forward.

Reply:
Okay, I added a TODO for `set_new_code_should_work`

Reply:
Think about it again, it's probably better to revert it for clarity and keep it ignored IMO as now it's harder to tell the purpose of this test, and doesn't test what we really want. Before, it was obvious that we would like to have a new primary runtime upgrade containing a new system domain runtime, and then the domain runtime is upgraded accordingly (`assert_eq!(runtime_code, new_runtime_wasm_blob);` means the latest domain runtime code must be updated to the one expected.) The domain runtime code remains the same in fact as we don't update it actually. In short, simply update the API `run_mock_primary_node` and keep it ignored, and then we can move forward to the other tests and merge this PR.

I also realize that we should improve the logic of checking whether a primary runtime is updated in case `RuntimeEnvironmentUpdated` will be associated with more events, we should check the corresponding event `CodeUpdated` directly instead.

Reply:
Okay, I'm good with both paths.

Reply:
The test has been reverted, plz take a look at the last commit.